proposals/span-safety.md

```diff
@@ -0,0 +1,188 @@
+# Compile time enforcement of safety for ref-like types #
+
+The main reason for the additional safety rules when dealing with types like `Span<T>` and `ReadonlySpan<T>` is that such types must be confined to the execution stack.
+ 
+There are two reasons why `Span<T>` and similar types must be a stack-only types.
+
+1. `Span<T>` is semantically a struct containing a reference and a range - `(ref T data, int length)`. Regardless of actual implementation, writes to such struct would not be atomic. Concurrent "tearing" of such struct would lead to the possibility of `length` not matching the `data`, causing out-of-range accesses and type-safety violations, which ultimately could result in GC heap corruption in seemingly "safe" code.
+2. Some implementations of `Span<T>` literally contain a managed pointer in one of its fields. Managed pointers are not supported as fields of heap objects and code that manages to put a managed pointer on the GC heap typically crashes at JIT time.
+
+All the above problems would be alleviated if instances of `Span<T>` are constrained to exist only on the execution stack. 
+
+An additional problem arises due to composition. It would be generally desirable to build more complex data types that would embed `Span<T>` and `ReadonlySpan<T>` instances. Such composite types would have to be structs and would share all the hazards and requirements of `Span<T>`. As a result the safety rules described here should be viewed as applicable to the whole range of **_ref-like types_**
+
+## `ref-like` types must be stack-only. ##
+
+C# compiler already has a concept of types with stack-only requirements that covers special platform types such as `TypedReference`. In some cases those types are referred as ref-like types as well. We should probably use some different term to refer to `TypedReference` and similar types - like `restricted types` to avoid confusion.
+
+In this document the "ref-like" types include only `Span<T>` and related types. And the ref-like variables mean variables of "ref-like" types.  
+
+In order to force ref-like variables to be stack only, we need the following restrictions: 
+
+- ref-like type cannot be a type of an array element
+- ref-like type cannot be used as a generic type argument
+- ref-like variable cannot be boxed
+- ref-like type cannot be a static field
+- ref-like type cannot be an instance field of ordinary not ref-like type
+- indirect restrictions, such as disallowed use of ref-like typed parameters and locals in async methods, which are really a result of disallowing ref-like typed fields.
+
+Note:
+
+- it is ok to return ref-like values from methods/lambdas, including by reference.
+- it is ok to pass ref-like values to methods/lambdas, including by reference.
+- it is ok for a ref-like type be a field of another struct as long as the struct itself is ref-like.
+- it is ok for an intermediate span values to be used in lambdas/async methods. 
```

Comment:
This needs explaining - you've not explained "intermediate span values" previously.